### PR TITLE
Remove globals in SpirvLowerRayTracing and move it after SpirvLowerGlobal

### PR DESCRIPTION
**Pull request by LLJJDD**
This PR contains a bunch of commits that gradually remove global variables used in SpirvLowerRayTracing pass and finally move it after SpirvLowerGlobal pass.

See commit message in each for detail.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 2663b21d4e306decb5b76e916ba0f34bd7a5b885
<details><summary>CTS tests (Failed: 38/211879)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35146/70631 (49.8%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35485/70631 (50.2%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39006/70632 (55.2%)</li>
<li><details><summary>Failed: 19/70632 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.ahit_triangles_256_256
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.chit_aabs_256_256
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.chit_aabs_32_32
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.chit_aabs_64_64
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.chit_triangles_32_32
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.sect_aabs_32_32
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TL1BL1G_AABB
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TL1BL1G_tri
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TL1BLnG_AABB
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TL1BLnG_tri
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TLnBL1G_AABB
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TLnBL1G_tri
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TLnBLnG_AABB
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TLnBLnG_tri
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.1024
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.16
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.256
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.4
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.64
Stack trace: Crash
```
</p></details></li>

<li>Not Supported: 31607/70632 (44.7%)</li><li>Warnings: 0/70632 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39038/70616 (55.3%)</li>
<li><details><summary>Failed: 19/70616 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.ahit_triangles_256_256
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.chit_aabs_256_256
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.chit_aabs_32_32
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.chit_aabs_64_64
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.chit_triangles_32_32
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.builtin.geometryindexext.sect_aabs_32_32
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TL1BL1G_AABB
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TL1BL1G_tri
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TL1BLnG_AABB
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TL1BLnG_tri
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TLnBL1G_AABB
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TLnBL1G_tri
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TLnBLnG_AABB
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.misc.NO_DUPLICATE_ANY_HIT_1TLnBLnG_tri
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.1024
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.16
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.256
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.4
Stack trace: Crash

FAILURE: dEQP-VK.ray_tracing_pipeline.watertightness.closedFan2.64
Stack trace: Crash
```
</p></details></li>

<li>Not Supported: 31559/70616 (44.7%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
Remember to remove the SpirvLowerRayTracingBuiltin source&header files.

The first two commits seem fine to me except for comments.

For the PR overall, I may have to look deeper still, but I don't think I agree with most of the dialect changes. It shouldn't be necessary to have all the `.ptr` versions

File: shared/lgcrt/include/lgcrt/LgcRtDialect.td [hunk 1]
```diff
@@ -272,6 +275,33 @@ def ReportHitOp : LgcRtOp<"report.hit", []> {
...
+  let description = [{
+    Return the current shader index.
+
+    The shader index is used to index into the shader binding table, i.e. the table which backs the
+    Vulkan shader record buffers or the DX12 local root signatures.
+  }];
+}
+
+// =========================================================================================================
+def GetShaderRecordBufferPtrOp : LgcRtOp<"get.shader.record.buffer.ptr", [Memory<[]>, WillReturn]> {
```
Comment (nhaehnle) [inline shared/lgcrt/include/lgcrt/LgcRtDialect.td:293]: Rename to just ShaderRecordBuffer / shader.record.buffer

File: lgc/interface/lgc/GpurtDialect.td [hunk 2]
```diff
@@ -304,3 +304,9 @@ def GpurtGetRayStaticIdOp : GpurtOp<"get.ray.static.id", [Memory<[(read Inaccess
   let results = (outs I32:$result);
   let summary = "get current ray static ID";
 }
+
+def GpurtGetGlobalPayloadPtrOp : GpurtOp<"get.global.payload.ptr", [Memory<[]>, WillReturn]> {
```
Comment (nhaehnle) [inline lgc/interface/lgc/GpurtDialect.td:308]: It's not clear to me why we would need this. I thought we replace SPIR-V payload globals by allocas and then pass pointers to them to all the relevant lgc.rt ops (e.g., `lgc.rt.call.callable.shader` has an explicit `param` argument).

**Review by nhaehnle — COMMENTED** [review]
On further inspection, I don't understand the justification for any of the *PtrOps for raytracing builtins, except perhaps RayTcurrent.

RayTcurrent has the fundamental issue that it is considered an Input in Vulkan/SPIR-V, but it's "volatile" in the sense that it may change during the execution of an intersection shader. But RayTcurrent is actually not the only builtin with this issue, because also SubgroupLocalInvocationIndex (and some related builtins) can change across invocation reorder points (like OpTraceRay), at least in principle.

So, since this is an issue that affects not just raytracing builtins (though it is restricted to raytracing shaders), how about addressing this in SpirvLowerGlobal as follows:

* Create a proxy alloca like traditional
* Create an `lgc.ray.tcurrent` and store into the proxy at function entry
* Find all points where the builtin may change, and repeat the sequence of `lgc.ray.tcurrent` + store to proxy in all those locations

The main motivation here is to keep all SPIR-V specific handling in SpirvLowerGlobal, as opposed to SpirvLowerRaytracing (which will be replaced at least in some cases by a generic pass).

**Comment by amdvlk-admin** [comment]
### Test summary for commit 2b400ebb84e5d839c05530aa5b9c8082d94f410d
<details><summary>CTS tests (Failed: 0/211879)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35146/70631 (49.8%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35485/70631 (50.2%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39025/70632 (55.3%)</li>
<li>Failed: 0/70632 (0.0%)</li>
<li>Not Supported: 31607/70632 (44.7%)</li><li>Warnings: 0/70632 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39057/70616 (55.3%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31559/70616 (44.7%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/interface/lgc/GpurtDialect.td (hunk 2 above)
Comment (LLJJDD) [reply lgc/interface/lgc/GpurtDialect.td:308]: This is needed especially for `TraceRay()` from GPURT, which we do not have the control of its parameter. But it seems that it is no longer needed after #2612 .

**Comment by LLJJDD** [comment]
> On further inspection, I don't understand the justification for any of the *PtrOps for raytracing builtins, except perhaps RayTcurrent.
> 

Yeah, the story behind is that I was using non-ptr dialect at first, and encounter problems with CTS using RayTcurrent, so I added the ptr version of dialect, and for consistency (and to match previous behavior), I replace all similar use cases.

> So, since this is an issue that affects not just raytracing builtins (though it is restricted to raytracing shaders), how about addressing this in SpirvLowerGlobal as follows:
> 
> * Create a proxy alloca like traditional
> * Create an `lgc.ray.tcurrent` and store into the proxy at function entry
> * Find all points where the builtin may change, and repeat the sequence of `lgc.ray.tcurrent` + store to proxy in all those locations
> 

This does sound good, thanks! I will address it in the next iteration. (I'm OOO today, so the change won't happen until next week.)

**Comment by LLJJDD** [comment]
I did a rebase so the compare button might be useless. Please look at the last commit to see what is actually changed.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7c5a52646e52aea80398b9292213a92ff4b4134d
<details><summary>CTS tests (Failed: 6/211876)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 8.9, Gfx10</summary><ul>

<li>Passed: 35497/70631 (50.3%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35134/70631 (49.7%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39377/70629 (55.8%)</li>
<li><details><summary>Failed: 2/70629 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.binding_model.descriptorset_random.sets32.dynindexed.ubolimitlow.sbolimitlow.sampledimglow.storageimghigh.noiub.uab.rgen.noia.0
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.descriptorset_random.sets4.dynindexed.ubolimitlow.nosbo.nosampledimg.outimgonly.noiub.nouab.rgen.noia.5
Stack trace: Crash
```
</p></details></li>

<li>Not Supported: 31250/70629 (44.2%)</li><li>Warnings: 0/70629 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39409/70616 (55.8%)</li>
<li><details><summary>Failed: 4/70616 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.binding_model.descriptorset_random.sets16.runtimesize.ubolimithigh.sbolimitlow.sampledimglow.outimgtexlow.iublimitlow.uab.chit.noia.0
Stack trace: Crash

FAILURE: dEQP-VK.binding_model.descriptorset_random.sets32.dynindexed.ubolimithigh.nosbo.nosampledimg.outimgonly.iublimithigh.uab.sect.noia.0
Stack trace: Crash

FAILURE: dEQP-VK.binding_model.descriptorset_random.sets4.dynindexed.ubolimithigh.nosbo.nosampledimg.outimgonly.iublimitlow.uab.chit.noia.0
Stack trace: Crash

FAILURE: dEQP-VK.binding_model.descriptorset_random.sets4.runtimesize.ubolimitlow.sbolimitlow.sampledimglow.lowimgnotex.noiub.nouab.rgen.noia.9
Stack trace: Crash
```
</p></details></li>

<li>Not Supported: 31203/70616 (44.2%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/lower/llpcSpirvLowerGlobal.cpp [hunk 3]
```diff
@@ -2472,4 +2472,61 @@ void SpirvLowerGlobal::lowerShaderRecordBuffer() {
...
+      Value *proxy;
+    };
+    Payload payload = {m_builder, proxy};
+
+    // RayTcurrent may change after OpReportIntersectionKHR, get and store it to proxy again after each Op is called.
+    static auto reportHitVisitor = llvm_dialects::VisitorBuilder<Payload>()
+                                       .setStrategy(llvm_dialects::VisitorStrategy::ByFunctionDeclaration)
+                                       .add<ReportHitOp>([](auto &payload, auto &op) {
+                                         payload.builder->SetInsertPoint(op.getNextNonDebugInstruction());
+                                         auto newRayTCurrent = payload.builder->template create<RayTcurrentOp>();
```
Comment (nhaehnle) [inline llpc/lower/llpcSpirvLowerGlobal.cpp:2517]: Nice, you don't often see this use of the template keyword in the wild ;)

**Review by nhaehnle — APPROVED** [review]
Nice. LGTM, maybe the test failures are spurious?

By the way, I wrote https://git.sr.ht/~nhaehnle/diff-modulo-base precisely to help with this problem of rebases ;)

**Comment by LLJJDD** [comment]
> Nice. LGTM, maybe the test failures are spurious?

Unfortunately the failures are valid.

> By the way, I wrote https://git.sr.ht/~nhaehnle/diff-modulo-base precisely to help with this problem of rebases ;)

Wow, that's cool!

**Comment by amdvlk-admin** [comment]
### Test summary for commit 6511a03e5b03341f8c79b8748d7e760863faa066
<details><summary>CTS tests (Failed: 0/211876)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35496/70631 (50.3%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35135/70631 (49.7%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39379/70629 (55.8%)</li>
<li>Failed: 0/70629 (0.0%)</li>
<li>Not Supported: 31250/70629 (44.2%)</li><li>Warnings: 0/70629 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39413/70616 (55.8%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31203/70616 (44.2%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]

**Pull request merged**